### PR TITLE
STYLE: Enable ruff TCH on "pandas/_libs/*"

### PR DESCRIPTION
**Pull request by v-mcoutinho**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

#51740, for
```
"pandas/_libs/*"
```

### CODE REVIEW

File: pandas/_libs/index.pyi [hunk 1]
```diff
@@ -1,9 +1,11 @@
-import numpy as np
+from typing import TYPE_CHECKING
 
-from pandas._typing import npt
+if TYPE_CHECKING:
```
Comment (twoertwein) [inline pandas/_libs/index.pyi:3]: If ruff requires this change within a stub file, please report an issue at ruff. Using `TYPE_CHECKING` inside a pyi files doesn't make sense as it is only used by type checkers.

Comment (MarcoGorelli) [reply pandas/_libs/index.pyi:3]: good point, reported

I'm still seeing it in 0.0.254 though:
```console
(.311venv) marcogorelli@DESKTOP-U8OKFP3:~/pandas-dev$ git diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 484107af67..ba57fcb432 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -28,7 +28,7 @@ repos:
         types_or: [python, pyi]
         additional_dependencies: [black==23.1.0]
 -   repo: https://github.com/charliermarsh/ruff-pre-commit
-    rev: v0.0.253
+    rev: v0.0.254
     hooks:
     -   id: ruff
 -   repo: https://github.com/jendrikseipp/vulture
diff --git a/pyproject.toml b/pyproject.toml
index 58a70b584c..f2f8ed9093 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -311,7 +311,6 @@ exclude = [
 "pandas/tests/*" = ["TCH"]
 "pandas/plotting/*" = ["TCH"]
 "pandas/util/*" = ["TCH"]
-"pandas/_libs/*" = ["TCH"]
 # Keep this one enabled
 "pandas/_typing.py" = ["TCH"]
 
(.311venv) marcogorelli@DESKTOP-U8OKFP3:~/pandas-dev$ pre-commit run ruff -a | grep 'index'
pandas/_libs/index.pyi:1:8: TCH002 Move third-party import `numpy` into a type-checking block
pandas/_libs/index.pyi:3:28: TCH001 Move application import `pandas._typing.npt` into a type-checking block
pandas/_libs/index.pyi:5:20: TCH001 Move application import `pandas.MultiIndex` into a type-checking block
pandas/_libs/index.pyi:6:32: TCH001 Move application import `pandas.core.arrays.ExtensionArray` into a type-checking block
pandas/_libs/window/indexers.pyi:1:8: TCH002 Move third-party import `numpy` into a type-checking block
pandas/_libs/window/indexers.pyi:3:28: TCH001 Move application import `pandas._typing.npt` into a type-checking block
```

Comment (MarcoGorelli) [reply pandas/_libs/index.pyi:3]: looks like they've fixed it https://github.com/charliermarsh/ruff/pull/3362/files

temporarily marking this PR as 'draft' until the next ruff version comes out, hope that's OK

Comment (twoertwein) [reply pandas/_libs/index.pyi:3]: Or could also have something like

```toml
[tool.ruff.per-file-ignores]
"*.pyi" = ["TCH"]  # TODO: remove line when using ruff 0.0.255
```

in pyproject.toml

Comment (MarcoGorelli) [reply pandas/_libs/index.pyi:3]: looks like there's a new ruff version

@v-mcoutinho do you want to update the ruff version please, revert the other changes, and remove `pandas/_libs` from pyproject.toml?

Comment (v-mcoutinho) [reply pandas/_libs/index.pyi:3]: I can't seem to find this new version, where are you seeing it exactly? ruff's github page still shows 0.0.254 as the latest version.

Comment (twoertwein) [reply pandas/_libs/index.pyi]: They just had a new release

Comment (v-mcoutinho) [reply pandas/_libs/index.pyi]: I made the changes, it seems the PR closed when I tried to push but I've opened it again.

**Review by twoertwein — APPROVED** [review]
LGTM, assuming the CI is green

**Review by MarcoGorelli — APPROVED** [review]
Nice one, thanks @v-mcoutinho (and @twoertwein of course!)

**Pull request merged**